# <font color="Blue">1990s Classic Hits Data  </font> 
* **Student: Lewis King**
* **ID: 000051992**

1990s Hits Dataset: ([Click here to see the 1990s Hits Dataset](https://www.kaggle.com/datasets/thebumpkin/1990s-classic-hits-with-spotify-data))

- Introduction: <br>
The 1990s was an important decade for music and many of people's favorite songs were released during that decade. Analyzing this data set will show which songs from the 1990s were the most popular, who were the artists that ran the decade, and ultimately which songs people enjoyed the most. This will be explored as the data set is cleaned and analyzed. 


In [145]:
#This code here imports the Libraries 
import pandas as pd
import numpy as np

In [146]:
#This will allow pandas to read and open the 1990s Hits Dataset
df= pd.read_csv('1990sClassicHits.csv')
df

Track                                Artist Duration  \
0        Everybody Everybody                             Black Box     4:11   
1               Close To You                           Maxi Priest     3:45   
2                     Do Me!                        Bell Biv DeVoe     3:56   
3                 Everything                           Jody Watley     4:05   
4              What It Takes                             Aerosmith     3:21   
..                       ...                                   ...      ...   
914           I’m Your Angel              R. Kelly and Celine Dion     3:12   
915        Taking Everything                         Gerald Levert     5:25   
916     What’s It Gonna Be?!          Busta Rhymes featuring Janet     2:41   
917     Almost Doesn’t Count                                Brandy     3:38   
918  It Ain’t My Fault 1 & 2  Silkk the Shocker Featuring Mystikal     3:30   

     Time_Signature  Danceability  Energy  Key  Loudness  Mode  Speechiness  \
0                 4         0.645   0.795    7   -12.095     1       0.0527   
1                 4         0.590   0.727    6    -7.955     1       0.0706   
2                 4         0.615   0.534    9    -6.719     0       0.0386   
3                 4         0.704   0.225    6   -14.454     0       0.0994   
4                 4         0.672   0.655   10    -5.021     0       0.0310   
..              ...           ...     ...  ...       ...   ...          ...   
914               4         0.648   0.331    4   -10.952     1       0.0274   
915               4         0.689   0.710    3    -5.395     0       0.0413   
916               4         0.710   0.742    9    -4.699     0       0.0339   
917               4         0.682   0.543    0    -7.063     0       0.0454   
918               4         0.743   0.567    4   -12.468     0       0.2570   

     Acousticness  Instrumentalness  Liveness  Valence    Tempo  Popularity  \
0          0.3470          0.003890    0.1040    0.535  112.067          87   
1          0.0259          0.000065    0.0827    0.356  121.966          85   
2          0.1060          0.000018    0.0607    0.193  135.917          84   
3          0.9020          0.657000    0.1060    0.243  120.006          82   
4          0.0362          0.000000    0.1170    0.555  134.948          82   
..            ...               ...       ...      ...      ...         ...   
914        0.7760          0.000445    0.1260    0.507   98.572          28   
915        0.0886          0.000000    0.1070    0.306  113.722          26   
916        0.1780          0.000064    0.0800    0.504  127.969          26   
917        0.4940          0.000000    0.1580    0.478   71.582          19   
918        0.1960          0.000000    0.2280    0.755   90.096           0   

     Year  
0    1990  
1    1990  
2    1990  
3    1990  
4    1990  
..    ...  
914  1999  
915  1999  
916  1999  
917  1999  
918  1999  

[919 rows x 17 columns]

## SECTION I: DATASET OVERVIEW AND STRUCTURE 

### 1. What is the shape and structure of the dataset?
- Use .shape, .info(), .head(), .tail()
- Show the first and last few rows of the dataset.
- Describe the general information of the DataFrame.

In [148]:
#This function shows the shape of the dataset 
df.shape 

(919, 17)

The output above shows that the shape of the dataset is 919 rows and 17 columns.

In [150]:
# This function is going to output the general information of the Dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 919 entries, 0 to 918
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Track             919 non-null    object 
 1   Artist            919 non-null    object 
 2   Duration          919 non-null    object 
 3   Time_Signature    919 non-null    int64  
 4   Danceability      919 non-null    float64
 5   Energy            919 non-null    float64
 6   Key               919 non-null    int64  
 7   Loudness          919 non-null    float64
 8   Mode              919 non-null    int64  
 9   Speechiness       919 non-null    float64
 10  Acousticness      919 non-null    float64
 11  Instrumentalness  919 non-null    float64
 12  Liveness          919 non-null    float64
 13  Valence           919 non-null    float64
 14  Tempo             919 non-null    float64
 15  Popularity        919 non-null    int64  
 16  Year              919 non-null    int64  
dt

The output above shows the general information of the data showing that there are 17 different columns,  null value count, and the data type. 

In [152]:
#This function is going to display the first 5 rows of the dataset.
df.head()

Track          Artist Duration  Time_Signature  Danceability  \
0  Everybody Everybody       Black Box     4:11               4         0.645   
1         Close To You     Maxi Priest     3:45               4         0.590   
2               Do Me!  Bell Biv DeVoe     3:56               4         0.615   
3           Everything     Jody Watley     4:05               4         0.704   
4        What It Takes       Aerosmith     3:21               4         0.672   

   Energy  Key  Loudness  Mode  Speechiness  Acousticness  Instrumentalness  \
0   0.795    7   -12.095     1       0.0527        0.3470          0.003890   
1   0.727    6    -7.955     1       0.0706        0.0259          0.000065   
2   0.534    9    -6.719     0       0.0386        0.1060          0.000018   
3   0.225    6   -14.454     0       0.0994        0.9020          0.657000   
4   0.655   10    -5.021     0       0.0310        0.0362          0.000000   

   Liveness  Valence    Tempo  Popularity  Year  
0    0.1040    0.535  112.067          87  1990  
1    0.0827    0.356  121.966          85  1990  
2    0.0607    0.193  135.917          84  1990  
3    0.1060    0.243  120.006          82  1990  
4    0.1170    0.555  134.948          82  1990

In [153]:
#This function is going to display the last 5 rows of the dataset.
df.tail()

Track                                Artist Duration  \
914           I’m Your Angel              R. Kelly and Celine Dion     3:12   
915        Taking Everything                         Gerald Levert     5:25   
916     What’s It Gonna Be?!          Busta Rhymes featuring Janet     2:41   
917     Almost Doesn’t Count                                Brandy     3:38   
918  It Ain’t My Fault 1 & 2  Silkk the Shocker Featuring Mystikal     3:30   

     Time_Signature  Danceability  Energy  Key  Loudness  Mode  Speechiness  \
914               4         0.648   0.331    4   -10.952     1       0.0274   
915               4         0.689   0.710    3    -5.395     0       0.0413   
916               4         0.710   0.742    9    -4.699     0       0.0339   
917               4         0.682   0.543    0    -7.063     0       0.0454   
918               4         0.743   0.567    4   -12.468     0       0.2570   

     Acousticness  Instrumentalness  Liveness  Valence    Tempo  Popularity  \
914        0.7760          0.000445     0.126    0.507   98.572          28   
915        0.0886          0.000000     0.107    0.306  113.722          26   
916        0.1780          0.000064     0.080    0.504  127.969          26   
917        0.4940          0.000000     0.158    0.478   71.582          19   
918        0.1960          0.000000     0.228    0.755   90.096           0   

     Year  
914  1999  
915  1999  
916  1999  
917  1999  
918  1999

### 2. Identify the variable types


In [155]:
#This code is going to identify the columns that include numerical data 
numerical_col = df.select_dtypes(include=['number']).columns.tolist()
#This code is going to identify the columns that include categorical data 
categorical_col = df.select_dtypes(include=['object']).columns.tolist()
print(f"Numerical Columns: {numerical_col}\n Categorical Columns: {categorical_col}")

Numerical Columns: ['Time_Signature', 'Danceability', 'Energy', 'Key', 'Loudness', 'Mode', 'Speechiness', 'Acousticness', 'Instrumentalness', 'Liveness', 'Valence', 'Tempo', 'Popularity', 'Year']
 Categorical Columns: ['Track', 'Artist', 'Duration']


As shown in the output above the variable type included in the dataset are categorical and numerical, which the columns showing categorical data being 'Track', 'Artist', and  'Duration', and the columns showing numerical data being 'Time_Signature', 'Danceability', 'Energy', 'Key', 'Loudness', 'Mode', 'Speechiness', 'Acousticness', 'Instrumentalness', 'Liveness', 'Valence', 'Tempo', 'Popularity', and 'Year'. 

### 3. Are there any missing values?
- How many missing values?
- Briefly describe how you plan to handle them.


In [158]:
df_new=df.copy()

In [159]:
# This code shows if there are null values in any of the columns and finds the sum of the null values per column
df_new.isnull().sum()

Track               0
Artist              0
Duration            0
Time_Signature      0
Danceability        0
Energy              0
Key                 0
Loudness            0
Mode                0
Speechiness         0
Acousticness        0
Instrumentalness    0
Liveness            0
Valence             0
Tempo               0
Popularity          0
Year                0
dtype: int64

The output above shows that there are no missing values to deal with in this dataset. However, if there were missing values I would first copy the dataset into a new file then using the new dataset I would replace the null values with the statement "Unknown" for a categorical column, and for numerical columns, I would find the average of the column and replace the missing values with the average. 

## SECTION II: DATA CLEANING & INDEXING 

### 4. Check for and remove duplicates
- Use .duplicated().sum() and .drop_duplicates() if applicable.


In [162]:
# This will check for duplicates and find the sum of the number of duplicates
duplicates= df_new.duplicated().sum()
print("Number of duplicate rows:", duplicates)

Number of duplicate rows: 0


As shown above, there are no duplicates in this dataset. However, if there were duplicates I would use the function .drop_duplivates() to remove all the duplicates.  

### 5. Provide two examples of DataFrame filtering using logical conditions
- Example: Filtering by category, threshold, etc.

In [165]:
# This filters songs numerically and by if they have a high popularity
popular = df_new.loc[df['Popularity'] > 80, 'Track'] 
#This filters songs categorically based on whether Madonna made them 
madonna_songs = df_new.loc[df['Artist'] == 'Madonna', 'Track']  
print("Songs with high popularity:\n", popular)
print("\nSongs made by Madonna:\n", madonna_songs)

Songs with high popularity:
 0         Everybody Everybody
1                Close To You
2                      Do Me!
3                  Everything
4               What It Takes
99                   Romantic
100                 Baby Baby
101        Losing My Religion
102            The First Time
199                   The One
200                       One
201                      Stay
202          Under The Bridge
203                 Tennessee
294                        If
295                      Rain
388                      Stay
389                    Linger
390                 If You Go
391                  The Sign
392                I Miss You
479           Someone To Love
480                   Believe
481                     Creep
482                      Baby
483                    I Know
484    Come And Get Your Love
565                    Nobody
566                      Time
567                       Hay
568                  Jealousy
643                   Someone
644        

Above the code provides a filtered list of song names included in the dataset that were either made by Maddonna or have a popularity of over 80. 

## SECTION III: DESCRIPTIVE STATISTICS 

### 6. Use .describe() on numerical columns. Share at least one interesting insight.

In [168]:
# This will display a list of all the numerical columns and provide statical data on them 
df_new.describe()

Time_Signature  Danceability      Energy         Key    Loudness  \
count      919.000000    919.000000  919.000000  919.000000  919.000000   
mean         3.947769      0.622967    0.623548    5.412405   -8.349071   
std          0.286769      0.154782    0.203975    3.514783    3.892821   
min          1.000000      0.135000    0.001340    0.000000  -42.950000   
25%          4.000000      0.518500    0.479500    2.000000  -10.166500   
50%          4.000000      0.631000    0.639000    6.000000   -7.599000   
75%          4.000000      0.738000    0.780000    9.000000   -5.751000   
max          5.000000      0.979000    0.999000   11.000000    0.919000   

             Mode  Speechiness  Acousticness  Instrumentalness    Liveness  \
count  919.000000   919.000000    919.000000        919.000000  919.000000   
mean     0.655060     0.075793      0.230068          0.052947    0.171865   
std      0.475608     0.079570      0.256284          0.180214    0.138296   
min      0.000000     0.023000      0.000004          0.000000    0.025000   
25%      0.000000     0.032400      0.029700          0.000000    0.089200   
50%      1.000000     0.042900      0.120000          0.000006    0.118000   
75%      1.000000     0.075600      0.344500          0.000741    0.217500   
max      1.000000     0.529000      0.995000          0.987000    0.965000   

          Valence       Tempo  Popularity         Year  
count  919.000000  919.000000  919.000000   919.000000  
mean     0.531424  118.711789   57.722524  1994.405876  
std      0.246091   29.684070   17.599522     2.926405  
min      0.006720   48.718000    0.000000  1990.000000  
25%      0.341500   96.924000   49.000000  1992.000000  
50%      0.535000  115.010000   60.000000  1994.000000  
75%      0.728500  133.424000   70.000000  1997.000000  
max      0.975000  219.973000   92.000000  1999.000000

Something interesting about this is that on average most of these sounds were released in 1994. 

### 7. Use .value_counts() to analyze at least two categorical columns.


In [171]:
#This code will display the number of times an Artist apperes int he dataset and take the top 5 common artists found in the dataset
df_new['Artist'].value_counts().head()

Artist
Mariah Carey       16
Madonna            14
Janet Jackson      13
Boyz II Men        11
Whitney Houston    10
Name: count, dtype: int64

As seen above, the most common artists to appear on the data set are Mariah Carey, Madonna, Janet Jackson, Boyz II Men, and Whitney Houston. Mariah Carey had the most hits on this list with 16. 

In [173]:
#This will find the most common frequencies in song duration and display the most common song length 
df_new['Duration'].value_counts()

Duration
3:55    12
3:48    12
3:59    11
4:18    11
4:26    10
        ..
3:04     1
5:08     1
5:47     1
3:06     1
4:46     1
Name: count, Length: 253, dtype: int64

As seen above the most common song lengths are 3:55 and 3:48 while the least common are 3:04, 5:08, 5:47, 3:06, and 4:46. 

### 8. Use .groupby() to compute summary statistics
- Example: Compute mean or count across categories.
- Include any additional interesting filtering/grouping insights.

In [176]:
#This will group by artists to find the average popularity of each artist's song and then display the top 10 airtist with the highest average populartiy for songs. 
df_new.groupby('Artist')['Popularity'].mean().nlargest(10)

Artist
Barenaked Ladies                     91.0
Jon B. featuring Babyface            87.0
SWV and Puff Daddy                   87.0
Brandy and Monica                    86.0
Faith Evans featuring Puff Daddy     86.0
Keith Sweat featuring Athena Cage    86.0
Smash Mouth                          86.0
Dionne Farris                        85.0
Erykah Badu                          85.0
Maxi Priest                          85.0
Name: Popularity, dtype: float64

The output above displays the top 10 artists with the highest average popularity, and we see that Barenaked Ladies has the highest popularity average on the list with 91.

## SECTION IV: OBSERVATIONS & QUESTIONS 

### 9. Highlight 1 key takeaways or patterns from your descriptive analysis.

One thing I noticed and thought was interesting was that even though Mariah Carey, Madonna, Janet Jackson, Boyz II Men, and Whitney Houston had the most songs in the dataset and are arguably the most popular artists in this dataset none of them made it to the average top popularity of tracks in the dataset. 

### 10. Suggest 2 questions to explore further in Part 2 using visualization.

- What common factors do the songs with a popularity over 80 have?
- What makes a song more danceable than another? 

## Conclusion 


To conclude, this dataset provided us with music hits from the 1990s, and throughout the data cleaning and analyzing we found which songs were the most popular which artists are the most common in this dataset, and the most common frequency of song duration. This dataset provided us with a good insight into the classic feeling of 1990s music and therefore allows us to connect with the past a little bit by analyzing the data of some of our favorite songs.  